### PR TITLE
feat(SessionSerializer): read ability for transitional serializer

### PR DESCRIPTION
**Pull request by maxiuyuan**
Problem: based on https://github.com/getsentry/sentry/issues/30861 we are trying to get rid of the PickleSerializer. 

Solution: As a transitional step into using the Json Serializer, a transitional serializer is made to read all pickle objects still. In preparation for reading json objects, it could read both json and pickle objects

#sync-getsentry

### CODE REVIEW

File: src/sentry/utils/transitional_serializer.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+    def dumps(self, obj: Dict[str, Any]) -> bytes:
+        # will have the write as pickle obj for now, will change to json in seperate deploy
+        # return self.json_serializer.dumps(obj)
+        return self.pickle_serializer.dumps(obj)
+
+    def loads(self, data: bytes) -> Dict[str, Any]:
+        try:
+            return self.json_serializer.loads(data)
+        except BaseException:
```
Comment (JoshFerge) [inline src/sentry/utils/transitional_serializer.py:20]: is there a more specific error we can except when the json serialization fails?

Comment (maxiuyuan) [reply src/sentry/utils/transitional_serializer.py:20]: definitely, i'll look into it and make the changes

File: tests/sentry/utils/test_transitional_serializer.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from django.contrib.sessions.serializers import PickleSerializer
+from django.core.signing import JSONSerializer
+from django.test.testcases import TestCase
+
+from sentry.utils.transitional_serializer import TransitionalSerializer
+
+
+class TransitionalSerializerTest(TestCase):
+    obj = {"id": 1, "user": "user", "email": "email"}
```
Comment (JoshFerge) [inline tests/sentry/utils/test_transitional_serializer.py:9]: probably worth having one more test that reads data that isn't json serializable -- like a datetime for example. this test can be removed in the next PR that changes writes.

**Comment by maxiuyuan** [comment]
#sync-getsentry

**Comment by maxiuyuan** [comment]
#sync-getsentry

File: src/sentry/utils/transitional_serializer.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        self.json_serializer = JSONSerializer()
+
+    def dumps(self, obj: Dict[str, Any]) -> bytes:
+        # will have the write as pickle obj for now, will change to json in seperate deploy
+        # return self.json_serializer.dumps(obj)
+        return self.pickle_serializer.dumps(obj)
+
+    def loads(self, data: bytes) -> Dict[str, Any]:
+        try:
+            metrics.incr("transitional_serializer_json_read")
```
Comment (JoshFerge) [inline src/sentry/utils/transitional_serializer.py:22]: I would have these be `transitional_serializer.json_read` (dot scopes it which may make datadog search a bit better). I would also add one metric for the picked write as well.

**Review by JoshFerge — APPROVED** [review]

**Pull request merged**